File: Hosts/Scala/ToGo/sanity/Main_test.scala
package sanity;

import openffi._;

class Main_test; object Main_test
{
	var t: Test = null;
	def main(args: Array[String])
	{
		try
		{
			t = new Test();

			testHelloWorld();
			testReturnsAnError();
			testDivIntegers();
			testJoinStrings();
		}
		catch
		{
			case e: Exception =>
				System.out.println("Unexpected exception has occurred during tests");
				e.printStackTrace();
				System.exit(1);
		}
	}

	def testHelloWorld()
	{
		t.HelloWorld();
	}

	def testReturnsAnError()
	{
		try
		{
			t.ReturnsAnError();
			System.out.println("Test should have failed");
			System.exit(1);
		}
		catch
		{
			case e: Exception =>
				if(!e.getMessage().equals("Panic in Go function. Panic Data: An error from ReturnsAnError"))
				{
					System.out.println("Unexpected error message.\nExpected: Panic in Go function. Panic Data: An error from ReturnsAnError\nReceived: "+e.getMessage());
					System.exit(1);
				}
		}
	}

	def testDivIntegers()
	{
		var res = t.DivIntegers(1, 2);
        if(res != 0.5)
        {
            System.out.println("Expected 0.5, got: "+res);
            System.exit(1);
        }

        try
        {
            t.DivIntegers(1, 0);
            System.out.println("Expected an error - divisor is 0");
            System.exit(1);
        }
        catch
        {
            case e: Exception =>
	            if(!e.getMessage().equals("Panic in Go function. Panic Data: Divisor is 0"))
	            {
	                System.out.println("Unexpected error message.\nExpected: Panic in Go function. Panic Data: Divisor is 0\nReceived: "+e.getMessage());
	                System.exit(1);
	            }
        }
	}

	def testJoinStrings()
	{
		var res = t.JoinStrings(Array("A","b","C"));
    	if(!res.equals("A,b,C"))
    	{
    	    System.out.println("Expected A,b,C. Got: "+res);
    	    System.exit(1);
    	}
	}
}

File: Hosts/Go/ToKotlin/sanity/Main_test.go
package sanity

import "testing"

//--------------------------------------------------------------------
func TestHelloWorld(t *testing.T){
	err := HelloWorld()
	if err != nil{
		t.Fatal(err)
	}
}
//--------------------------------------------------------------------
func TestReturnsAnError(t *testing.T){
	err := ReturnsAnError()
	if err == nil{
		t.Fatal("Error expected")
	}
}
//--------------------------------------------------------------------
func TestDivIntegers(t *testing.T){
	res, err := DivIntegers(1, 2)
	if err != nil{
		t.Fatal(err)
	}

	if res != 0.5{
		t.Fatalf("Expected 0.5, got: %v", res)
	}

	res, err = DivIntegers(1, 0)
	if err == nil{
		t.Fatal("Expected an error - divisor is 0")
	}
}
//--------------------------------------------------------------------
func TestJoinStrings(t *testing.T){
	res, err := JoinStrings([]string{"A", "b", "C"})
	if err != nil{
		t.Fatal(err)
	}

	if res != "A,b,C"{
		t.Fatalf("Expected A,b,C. Got: %v", res)
	}
}
//--------------------------------------------------------------------